simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp

```diff
@@ -1547,70 +1547,52 @@ auto HdMapUtils::getLongitudinalDistance(
   if (route.empty()) {
     return std::nullopt;
   }
-  double distance = 0;
-
-  auto with_lane_change = [this](
-                            const bool allow_lane_change, const lanelet::Id current_lanelet,
-                            const lanelet::Id next_lanelet) -> bool {
-    if (allow_lane_change) {
-      auto next_lanelet_ids = getNextLaneletIds(current_lanelet);
-      auto next_lanelet_itr = std::find_if(
-        next_lanelet_ids.begin(), next_lanelet_ids.end(),
-        [next_lanelet](const lanelet::Id & id) { return id == next_lanelet; });
-      return next_lanelet_itr == next_lanelet_ids.end();
-    } else {
-      return false;
-    }
+
+  const auto with_lane_change =
```

Comment:
check out something like this and let me know
```
auto HdMapUtils::getLongitudinalDistance(
  const traffic_simulator_msgs::msg::LaneletPose & from_pose,
  const traffic_simulator_msgs::msg::LaneletPose & to_pose, bool allow_lane_change) const
  -> std::optional<double>
{
  if (from_pose.lanelet_id == to_pose.lanelet_id) {
    if (from_pose.s > to_pose.s) {
      return std::nullopt;
    }
    return to_pose.s - from_pose.s;
  }

  const auto is_lane_change_required =
    [this](const lanelet::Id current_lanelet, const lanelet::Id next_lanelet) -> bool {
    const auto next_lanelet_ids = getNextLaneletIds(current_lanelet);
    return std::none_of(
      next_lanelet_ids.begin(), next_lanelet_ids.end(),
      [next_lanelet](const lanelet::Id id) { return id == next_lanelet; });
  };

  /// @sa https://tier4.github.io/scenario_simulator_v2-docs/developer_guide/DistanceCalculation/
  if (const auto route = getRoute(from_pose.lanelet_id, to_pose.lanelet_id, allow_lane_change);
      !route.empty()) {
    constexpr double matching_distance = 10.0;
    constexpr double lanelet_starting_s = 0.0;

    double accumulated_distance = 0.0;
    // sum the distance from the beginning of the current lanelet to the beginning of the last one
    for (std::size_t i = 0UL; i < route.size() - 1UL; ++i) {
      // if lane change is required, add the distance traveled during the lane change
      // if lane change is not required, add the total length of the current lanelet
      if (is_lane_change_required(route[i], route[i + 1UL])) {
        const auto current_lanelet_spline = getCenterPointsSpline(route[i]);
        const auto next_lanelet_spline = getCenterPointsSpline(route[i + 1UL]);

        const auto lanelet_half_min_s =
          std::min(current_lanelet_spline->getLength(), next_lanelet_spline->getLength()) * 0.5;

        const auto current_start_matching_s = current_lanelet_spline->getSValue(
          next_lanelet_spline->getPose(lanelet_starting_s), matching_distance);
        const auto next_start_matching_s = next_lanelet_spline->getSValue(
          current_lanelet_spline->getPose(lanelet_starting_s), matching_distance);
        const auto current_middle_matching_s = current_lanelet_spline->getSValue(
          next_lanelet_spline->getPose(lanelet_half_min_s), matching_distance);
        const auto next_middle_matching_s = next_lanelet_spline->getSValue(
          current_lanelet_spline->getPose(lanelet_half_min_s), matching_distance);

        if (current_start_matching_s.has_value()) {
          accumulated_distance += current_start_matching_s.value();
        } else if (next_start_matching_s.has_value()) {
          accumulated_distance -= next_start_matching_s.value();
        } else if (current_middle_matching_s.has_value()) {
          accumulated_distance += current_middle_matching_s.value() - lanelet_half_min_s;
        } else if (next_middle_matching_s.has_value()) {
          accumulated_distance -= next_middle_matching_s.value() - lanelet_half_min_s;
        } else {
          return std::nullopt;
        }
      } else {
        accumulated_distance += getLaneletLength(route[i]);
      }
    }
    // subtract the distance already traveled on the current lanelet: from_pose.s
    // and add the distance that needs to be traveled on the last: to_pose.s
    return accumulated_distance - from_pose.s + to_pose.s;
  } else {
    return std::nullopt;
  }
}
```

Reply:
these changes have been incorporated